libvast/src/system/index.cpp

```diff
@@ -831,38 +703,56 @@ void schedule_lookups(index_state& st) {
-      //   auto& query_state = st.pending_queries[qid];
-      //   query_state.candidate_partitions--;
-      // }
+      // We need to mark failed partitions as completed to avoid clients going
+      // out of sync.
+      for (auto qid : next->queries) {
+        auto it = st.pending_queries.queries().find(qid);
+        if (it == st.pending_queries.queries().end()) {
+          VAST_WARN("{} tried to access non-existant query {}", *st.self, qid);
+          continue;
+        }
+        auto& query_state = it->second;
+        query_state.completed_partitions++;
+        if (query_state.completed_partitions
+            == query_state.requested_partitions)
+          st.self->send(query_state.client, atom::done_v);
+        if (query_state.completed_partitions
+            == query_state.candidate_partitions)
+          st.pending_queries.queries().erase(qid);
+      }
       continue;
     }
-    st.counters.partition_materializations++;
+    st.counters.partition_schedulings++;
     st.counters.partition_lookups += next->queries.size();
     // 3. request all relevant queries in a loop
     auto cnt = std::make_shared<size_t>(next->queries.size());
     for (auto qid : next->queries) {
-      auto& query_state = st.pending_queries.queries[qid];
+      auto it = st.pending_queries.queries().find(qid);
+      if (it == st.pending_queries.queries().end()) {
+        VAST_WARN("{} tried to access non-existant query {}", *st.self, qid);
+        continue;
+      }
+      auto& query_state = it->second;
       auto handle_completion = [cnt, qid, &st] {
-        // TODO: use find
-        auto& query_state = st.pending_queries.queries[qid];
-        query_state.completed_partitions++;
-        if (query_state.completed_partitions
-            == query_state.requested_partitions) {
-          st.self->send(query_state.client, atom::done_v);
-          auto response = query_state.completed_partitions
-                              == query_state.candidate_partitions
-                            ? query_response::complete
-                            : query_response::incomplete;
-          query_state.rp.deliver(response);
+        auto it = st.pending_queries.queries().find(qid);
+        if (it == st.pending_queries.queries().end()) {
+          VAST_WARN("{} tried to access non-existant query {}", *st.self, qid);
+        } else {
+          auto& query_state = it->second;
+          query_state.completed_partitions++;
+          if (query_state.completed_partitions
+              == query_state.requested_partitions)
+            st.self->send(query_state.client, atom::done_v);
+          // 4. Remove queries with no more remaining partitions from
+          // pending_queries.
+          if (query_state.completed_partitions
+              == query_state.candidate_partitions) {
+            VAST_ASSERT(!st.pending_queries.reachable(qid));
+            st.pending_queries.queries().erase(qid);
+          }
```

Comment:
This should go into a separate function that takes the query id and returns whether you should send the `atom::done_v` to the client.

See also the above comment on removing the mutable `queries` overload.